lib/grp.gi

```diff
@@ -783,16 +805,35 @@ InstallMethod( DerivedSeriesOfGroup,
     D := DerivedSubgroup( G );
    
     while
-      # we don't know that the group has no generators
-      (not HasGeneratorsOfGroup(S[Length(S)]) or
-	    Length(GeneratorsOfGroup(S[Length(S)]))>0) and
-      ( (not HasAbelianInvariants(S[Length(S)]) and D <> S[ Length(S) ]) or
-	    Length(AbelianInvariants(S[Length(S)]))>0) do
+      (not HasIsTrivial(S[Length(S)]) or
+	    not IsTrivial(S[Length(S)])) and
+      (
+        (not HasIsPerfectGroup(S[Length(S)]) and
+         not HasAbelianInvariants(S[Length(S)]) and D <> S[ Length(S) ]) or
+        (HasIsPerfectGroup(S[Length(S)]) and not IsPerfectGroup(S[Length(S)]))
+        or (HasAbelianInvariants(S[Length(S)])
+                            and Length(AbelianInvariants(S[Length(S)])) > 0)
+      ) do
```

Comment:
There is one thing I wonder about, though: What if the group `D` (resp. `S[Length(S)]` does not "know" it is trivial, but actually *is* trivial... don't we then end up with a list `S` which contains two copies of the trivial group at the end? Same for the group being perfect...
(This is pertinent to all variants of the code: the current one,  hungaborhorvath's version, and my proposed variant...)